src/coreComponents/physicsSolvers/multiphysics/CoupledSolver.hpp

```diff
@@ -0,0 +1,241 @@
+/*
+ * ------------------------------------------------------------------------------------------------------------
+ * SPDX-License-Identifier: LGPL-2.1-only
+ *
+ * Copyright (c) 2018-2020 Lawrence Livermore National Security LLC
+ * Copyright (c) 2018-2020 The Board of Trustees of the Leland Stanford Junior University
+ * Copyright (c) 2018-2020 TotalEnergies
+ * Copyright (c) 2019-     GEOSX Contributors
+ * All rights reserved
+ *
+ * See top level LICENSE, COPYRIGHT, CONTRIBUTORS, NOTICE, and ACKNOWLEDGEMENTS files for details.
+ * ------------------------------------------------------------------------------------------------------------
+ */
+
+/**
+ * @file CoupledSolver.hpp
+ *
+ */
+
+#ifndef GEOSX_PHYSICSSOLVERS_MULTIPHYSICS_COUPLEDSOLVER_HPP_
+#define GEOSX_PHYSICSSOLVERS_MULTIPHYSICS_COUPLEDSOLVER_HPP_
+
+#include "physicsSolvers/SolverBase.hpp"
+
+#include <tuple>
+
+namespace geosx
+{
+
+template< typename ... SOLVERS >
+class CoupledSolver : public SolverBase
+{
+
+public:
+
+  /**
+   * @brief main constructor for ManagedGroup Objects
+   * @param name the name of this instantiation of ManagedGroup in the repository
+   * @param parent the parent group of this instantiation of ManagedGroup
+   */
+  CoupledSolver( const string & name,
+                 Group * const parent )
+    : SolverBase( name, parent )
+  {}
+
+
+  /// deleted copy constructor
+  CoupledSolver( CoupledSolver const & ) = delete;
+
+  /// default move constructor
+  CoupledSolver( CoupledSolver && ) = default;
+
+  /// deleted assignment operator
+  CoupledSolver & operator=( CoupledSolver const & ) = delete;
+
+  /// deleted move operator
+  CoupledSolver & operator=( CoupledSolver && ) = delete;
+
+  /**
+   * @defgroup Solver Interface Functions
+   *
+   * These functions provide the primary interface that is required for derived classes
+   */
+  /**@{*/
+
+  virtual void
+  setupDofs( DomainPartition const & domain,
+             DofManager & dofManager ) const override
+  {
+    forEachArgInTuple( std::tuple< SOLVERS *... >{}, [&]( auto t, auto idx )
+    {
+      GEOSX_UNUSED_VAR( t );
+      auto & solver = std::get< idx() >( m_solvers );
+      solver->setupDofs( domain, dofManager );
+    } );
```

Comment:
This will set up all Dofs for all solverts but not the coupling blocks. For some solvers, like https://github.com/GEOSX/GEOSX/blob/bb16d72e8dc43a94f3f55e9b6aa6dc7de003f669/src/coreComponents/physicsSolvers/multiphysics/SinglePhasePoromechanicsSolver.cpp#L121 the `setupDofs` call also adds coupling terms across variables of the two solvers. Would it make sense to make this function non-virtual and add another virtual function, e.g., `setupCoupling` and have solvers overload that one?

Reply:
Yes, you are right, I implemented what you suggested